README.md
# Mudkip

[![GitHub Actions](https://github.com/vberlier/mudkip/workflows/CI/badge.svg)](https://github.com/vberlier/mudkip/actions)
[![PyPI](https://img.shields.io/pypi/v/mudkip.svg)](https://pypi.org/project/mudkip/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/mudkip.svg)](https://pypi.org/project/mudkip/)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)

> A friendly [Sphinx](https://www.sphinx-doc.org/en/master/) wrapper.

Mudkip is a small wrapper around [Sphinx](https://www.sphinx-doc.org/en/master/) that bundles essential tools and extensions, providing everything needed for building rich documentation for Python projects.

```bash
$ mudkip --help
Usage: mudkip [OPTIONS] COMMAND [ARGS]...

  A friendly Sphinx wrapper.

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
  build    Build documentation.
  clean    Remove output directory.
  develop  Start development server.
  init     Initialize documentation.
  test     Test documentation.
```

## Features

Mudkip intends to provide an out-of-the-box solution for most Python projects. The command-line utility lets you build and check your documentation, launch a development server with live reloading, run doctests and more!

Mudkip enables the following Sphinx extensions:

- [`sphinx.ext.autodoc`](http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) for generating documentation from docstrings
- [`sphinx.ext.napoleon`](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html) for Google-style and NumPy-style docstring support
- [`sphinx.ext.doctest`](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) for doctest support
- [`sphinx.ext.autosectionlabel`](https://www.sphinx-doc.org/en/master/usage/extensions/autosectionlabel.html) for referencing sections with their title
- [`sphinx.ext.intersphinx`](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) for linking to other projects’ documentation
- [`sphinx.ext.githubpages`](https://www.sphinx-doc.org/en/master/usage/extensions/githubpages.html) when deploying to GitHub Pages
- [`myst_parser`](https://myst-parser.readthedocs.io/en/latest/) for markdown support
- [`myst_nb`](https://myst-nb.readthedocs.io/en/latest/) for Jupyter notebook support
- [`sphinxcontrib.mermaid`](https://github.com/mgaitan/sphinxcontrib-mermaid) for [Mermaid](https://mermaid-js.github.io/mermaid/) graphs and flowcharts

## Installation

The package can be installed with `pip`.

```bash
$ pip install mudkip
```

To use the development version:

```bash
$ pip install update git+https://github.com/vberlier/mudkip.git
```

## Getting started

You can forget everything about `sphinx-quickstart`, `conf.py` and intimidating Makefiles. After installing the package, no need to configure anything you can run the `develop` command right away and start writing docs.

```bash
$ mudkip develop
Watching "docs"...
Server running on http://localhost:5500
```

The command will create the `docs` directory and `index.rst` file if they do not already exist and launch a development server with live reloading. You can open the link in your browser and see that mudkip uses the [Read the Docs](https://github.com/rtfd/sphinx_rtd_theme) theme by default.

> Note that mudkip enables the [`myst_parser`](https://myst-parser.readthedocs.io/en/latest/) extension, allowing you to use both reStructuredText and markdown files. You can create an `index.md` file if you want to use markdown instead of reStructuredText.
> Check out this project's own [`docs` folder](https://github.com/vberlier/mudkip/tree/main/docs) as an example.

Press `Ctrl+C` at any time to exit.

Most changes in `index.rst` or other documentation source files will be processed immediately and shown in browser. Deep changes (eg using a different theme) take effect after you stop and restart the live server.

### Building and checking documentation

The `build` command invokes the [`dirhtml`](https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.dirhtml.DirectoryHTMLBuilder) builder and builds your documentation. By default, the generated files are in "docs/\_build".

```bash
$ mudkip build
```

Running the command with the `--check` or `-c` flag will exit with code `1` if Sphinx reports any error or warning.

```bash
$ mudkip build --check
```

The `--check` flag also makes sure that there are no broken links by running the [`linkcheck`](https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.linkcheck.CheckExternalLinksBuilder) builder on your documentation. You can disable this with the `--skip-broken-links` flag.

The `build` command also features a really handy flag if you're deploying the documentation to GitHub Pages. The `--update-gh-pages` flag will invoke Sphinx with the [`sphinx.ext.githubpages`](https://www.sphinx-doc.org/en/master/usage/extensions/githubpages.html) extension and then force push the output directory to the `gh-pages` branch of your repository.

```bash
$ mudkip build --update-gh-pages
```

The remote branch will be created if it doesn't already exist.

### Running doctests

Mudkip enables the [`sphinx.ext.doctest`](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) extension, making it possible to test interactive code examples. You can try it out by adding the following code snippet to your `index` document:

```rst
.. doctest::

    >>> import this
    The Zen of Python, by Tim Peters
    <BLANKLINE>
    Beautiful is better than ugly.
    ...
```

The `test` command will run the code example and make sure that the output matches the documentation.

```bash
$ mudkip test
Testing "docs"...

Document: index
---------------
1 items passed all tests:
   1 tests in default
1 tests in 1 items.
1 passed and 0 failed.
Test passed.

Doctest summary
===============
    1 test
    0 failures in tests
    0 failures in setup code
    0 failures in cleanup code

Passed.
```

### Using Jupyter notebooks

The [`myst-nb`](https://myst-nb.readthedocs.io/en/latest/) extension provides support for Jupyter notebooks. This means that in addition to `.rst` and `.md` files, Sphinx will also generate pages for `.ipynb` files.

The `develop` command can launch the Jupyter notebook in the "docs" directory and open it in your browser with the `--notebook` or `-n` flag.

```bash
$ mudkip develop --notebook
Watching "docs"...
Server running on http://localhost:5500
Notebook running on http://localhost:8888/?token=5e64df6...
```

Notebooks are executed during the build process. The `--check` flag will make sure that there are no uncaught exceptions in any cell.

### Integration with npm and yarn

Mudkip can help you go beyond traditional Sphinx themes by running npm scripts for you and integrate with the build process of a custom front-end. If your docs contain a `package.json` file, Mudkip will run Sphinx and then invoke the appropriate npm script using your preferred npm client.

```bash
$ mudkip build
```

Here, Mudkip would try to run either `npm run build` or `yarn build` before exiting the command. Similarly, `mudkip clean` would try to run either `npm run clean` or `yarn clean`.

```bash
$ mudkip develop
```

The `develop` command will try to run one of the following scripts: `develop`, `dev`, `start` or `serve`. If you don't have a dedicated script to run your project in development mode, Mudkip will simply execute the `build` script after running Sphinx each time you make a modification.

## Configuration

Mudkip doesn't really require any configuration but you can change some of the default settings with command-line options or a configuration file.

For example, when running a command, you can set the `--preset` or `-p` option to `furo` if you want to use the [Furo](https://pradyunsg.me/furo/) theme instead of the default [Read the Docs](https://github.com/rtfd/sphinx_rtd_theme) theme.

```
$ mudkip build --preset furo
```

It's also possible to change the default source and output directories with the `--source-dir` and `--output-dir` options respectively.

```
$ mudkip build --source-dir path/to/docs --output-dir path/to/output
```

Passing these options to every single command can become tedious so you can use a configuration file to save your custom settings.

Running the `init` command will either add a `[tool.mudkip]` section to your existing `pyproject.toml` or create a new `mudkip.toml` file with some basic configuration.

```bash
$ mudkip init
```

### Available options

- `preset`

  **default**: `"rtd"`

  Presets configure Mudkip and Sphinx to enable specific features. The `rtd`, `furo`, `pydata` and `alabaster` presets enable the development server and configure Sphinx to use the `dirhtml` builder. The `rtd` preset changes the html theme to the [Read the Docs](https://github.com/rtfd/sphinx_rtd_theme) theme, `furo` preset uses [Furo](https://pradyunsg.me/furo/) theme, and `pydata` changes to [PyData](https://pydata-sphinx-theme.readthedocs.io/en/stable/) theme.

  The `xml` preset configures Sphinx to use the `xml` builder. This is useful for more advanced use cases when you process a hierarchy of docutils documents further in your static site generator pipeline (experimental).

  The `latex` preset uses `latex` builder, which creates LaTeX source files for making a PDF version of your documentation. You may want to change `--output-dir` manually to a different directory when using `latex` preset to separate HTML and LaTeX outputs.

- `source_dir`

  **default**: `"docs"`

  This is the directory containing the source files for your documentation. Sphinx is configured to use it as its source directory and when the development server is enabled, Mudkip will watch the directory for changes to rebuild your documentation.

- `output_dir`

  **default**: `"docs/_build"`

  The output directory is where Sphinx will output the generated files. This is also the directory served by the development server.

- `base_url`

  **default**: `None`

  The base url used by Sphinx when building the documentation. You can use it to specify a custom domain when deploying to GitHub Pages and make sure Sphinx generates the appropriate `CNAME` file.

- `repository`

  **default**: The `repository` field in `pyproject.toml`

  The repository url of the remote when updating GitHub Pages.

  If you're not using [poetry](https://poetry.eustace.io/), you will need to set it manually.

- `verbose`

  **default**: `false`

  This option can also be enabled on the command-line with the `--verbose` flag. Setting it to `true` will tell Mudkip to display the entire Sphinx output as well as the Jupyter notebook output when running the `develop` command with the `--notebook` flag.

- `project_name`

  **default**: The name of the project you're documenting in `pyproject.toml`

  If you're not using [poetry](https://poetry.eustace.io/), you will need to set it manually.

- `project_dir`

  **default**: The value of the `project_name` option

  Mudkip will watch the Python files in your project directory when using the development server. This enables live reloading even when you're editing docstrings.

- `title`

  **default**: The value of the `project_name` option

  The project title used by Sphinx when building the documentation.

- `copyright`

  **default**: The current year followed by the value of the `author` option

  The copyright notice used by Sphinx when building the documentation.

- `author`

  **default**: The concatenated list of authors in `pyproject.toml`

  If you're not using [poetry](https://poetry.eustace.io/), you will need to set it manually.

- `version`

  **default**: The first two numbers of the `release` option

  The version used by Sphinx when building the documentation.

- `release`

  **default**: The project version in `pyproject.toml`

  If you're not using [poetry](https://poetry.eustace.io/), you will need to set it manually.

- `override`

  **default**: An empty dictionary

  The `override` option lets you specify sphinx configuration directly. For example, you can use it to define a custom theme or a logo image.

  Note that `override` option is effectively a replacement for `conf.py` configuration.
  You can put an `override` section in `mudkip.toml` file as shown in an example below.

  ```toml
  [mudkip.override]
  myst_enable_extensions = ["replacements", "tasklist"]
  html_static_path = ["_static"]
  ```

  In a `pyproject.toml` file the `override` section name should be `[tool.mudkip.override]`.

- `section_label_depth`

  Enables `sphinx.ext.autosectionlabel` and sets the `autosectionlabel_maxdepth` option.

## Contributing

Contributions are welcome. Make sure to first open an issue discussing the problem or the new feature before creating a pull request. The project uses [poetry](https://python-poetry.org/).

```bash
$ poetry install
```

The code follows the [black](https://github.com/psf/black) code style.

```bash
$ poetry run black mudkip
```

## Why package name Mudkip?

[wiki]: https://en.wikipedia.org/wiki/List_of_generation_III_Pok%C3%A9mon#Mudkip

This package started as a hobby project, which **@vberlier** usually named after Pokémon characters. [Mudkip][wiki] _"has a fin on its head, which acts as a radar for its surroundings. Even in muddy water, Mudkip can sense where it is going."_ Cute video about Mudkip [here](https://www.youtube.com/watch?v=l7kK9bxEGMg).

---

License - [MIT](https://github.com/vberlier/mudkip/blob/master/LICENSE)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.8.0 (2022-08-09)
### Feature
* Add latex preset ([`774f663`](https://github.com/vberlier/mudkip/commit/774f6637a2a4f653187901ca7d7d7fc600a212fc))
* Add preset for pydata theme ([`364be03`](https://github.com/vberlier/mudkip/commit/364be0378189d68bfd443c738cc2ae9e80f83f34))
* Add mermaid ([`2979d61`](https://github.com/vberlier/mudkip/commit/2979d6151a399ada8f3f83cc7e993592c2efec82))

## v0.7.0 (2022-05-31)
### Feature
* Remove mdinclude as the normal include directive supports custom parsers now ([`f7da3f0`](https://github.com/vberlier/mudkip/commit/f7da3f07b989d7476d234c49bbc8f0422faa7157))

## v0.6.1 (2022-05-31)
### Fix
* Change myst-nb settings ([`e6b2452`](https://github.com/vberlier/mudkip/commit/e6b245224e1761a5803f2b98d4ff03a3e2d108e1))
* Remove nb_render_priority ([`e703076`](https://github.com/vberlier/mudkip/commit/e703076323ae1cd7771e4a16b3841a3104b2b533))

## v0.6.0 (2022-05-31)
### Feature
* New myst-nb version that integrates recent markdown-it-py ([`6281675`](https://github.com/vberlier/mudkip/commit/62816755ab74f7adf4992da38dc05073cd0e4bc4))

## v0.5.3 (2022-04-05)
### Fix
* Relax myst-nb requirement ([`67f1ea9`](https://github.com/vberlier/mudkip/commit/67f1ea924f1f879cde2fb6d456eb58f77816c2bc))

## v0.5.2 (2022-04-05)
### Fix
* Update dependencies ([`e857c0a`](https://github.com/vberlier/mudkip/commit/e857c0ad90e513bb8d62485d84d7ccf06e31f7bf))

## v0.5.1 (2022-02-20)
### Fix
* Disable smartquotes by default ([`3822594`](https://github.com/vberlier/mudkip/commit/382259450b68d0b5fe51fb307abd76729ee314d8))

## v0.5.0 (2022-02-20)
### Feature
* Add wip vitepress preset ([`fc8c928`](https://github.com/vberlier/mudkip/commit/fc8c928821cebb07ce20da5eb2b9adfc8e724a92))

## v0.4.15 (2022-02-19)
### Fix
* Use native autodoc typehints and add intersphinx ([`ed51b54`](https://github.com/vberlier/mudkip/commit/ed51b5480e2b3d2735350fe69dba1ef285a465e7))

## v0.4.14 (2021-11-20)
### Fix
* Take into account the output directory when detecting changes ([`a5749f4`](https://github.com/vberlier/mudkip/commit/a5749f473064896dbde80a96eb366a5779e5b1b4))

## v0.4.13 (2021-11-20)
### Fix
* Tweak watch patterns ([`76f42b5`](https://github.com/vberlier/mudkip/commit/76f42b547970b43d17c3722f3234ebb57397dfea))

## v0.4.12 (2021-09-23)
### Fix
* Update furo to the latest version ([`0f0707a`](https://github.com/vberlier/mudkip/commit/0f0707af53de1328ef5b4228dd286042b055c1c7))

## v0.4.11 (2021-09-12)
### Fix
* Update dependencies ([`713520a`](https://github.com/vberlier/mudkip/commit/713520a39938c91275706b3236c9bf1eecbf0315))

## v0.4.10 (2021-08-15)
### Fix
* Upgrade to click 8 ([`40c0932`](https://github.com/vberlier/mudkip/commit/40c0932461bdbefe2f76494baa5f9f1a848335a7))

## v0.4.9 (2021-05-05)
### Fix
* Only open notebook in browser with the --open flag ([`25763c7`](https://github.com/vberlier/mudkip/commit/25763c7700be97db887ccc6d0c9d791d7af12584))

## v0.4.8 (2021-05-04)
### Fix
* Load extensions earlier so that overrides can change extension config ([`9fdfbd4`](https://github.com/vberlier/mudkip/commit/9fdfbd4739bd4b36cd53ed1a8694c1cf76d7e176))

## v0.4.7 (2021-05-04)
### Fix
* Update README ([`8bf7237`](https://github.com/vberlier/mudkip/commit/8bf72375a155cf0f437796bb9c577d708e897d09))

### Documentation
* Add changelog ([`462ae75`](https://github.com/vberlier/mudkip/commit/462ae7596949e1a4f7dd880ba14ed7aa50d66655))

## v0.4.6 (2021-05-04)
### Fix
* Improve furo cell outputs ([`f346fd8`](https://github.com/vberlier/mudkip/commit/f346fd8b4d4b4edbe7d2576cef50b270814a7114))

### Documentation
* Fix broken links ([`8f2ef32`](https://github.com/vberlier/mudkip/commit/8f2ef32c696728a11e4656ef21445c91723669f9))
* Update README ([`ec5a718`](https://github.com/vberlier/mudkip/commit/ec5a718c6f7f812e1c765e6f422a5db1f9396f29))
* Add homepage and example notebook ([`e79cbfe`](https://github.com/vberlier/mudkip/commit/e79cbfeb78d35ab28f5e1351c0a417c37182b003))


mudkip/cli.py
import sys
import time
from contextlib import contextmanager
from functools import wraps
from os import path
from traceback import format_exc, print_exception

import click

from . import __version__
from .application import Mudkip
from .config import Config
from .errors import MudkipError
from .preset import Preset

DIRECTORY = click.Path(file_okay=False)


@contextmanager
def exception_handler(exit=False, verbose=False):
    try:
        yield
    except Exception as exc:
        error = exc.args[0] if isinstance(exc, MudkipError) else format_exc()
        click.secho(error, fg="red", bold=True)

        if verbose and isinstance(exc, MudkipError) and exc.__cause__:
            exc = exc.__cause__
            click.echo()
            print_exception(type(exc), exc, exc.__traceback__)

        if exit:
            sys.exit(1)


def print_version(ctx, _param, value):
    if not value or ctx.resilient_parsing:
        return
    click.secho(f"Mudkip v{__version__}", fg="cyan")
    ctx.exit()


@click.group()
@click.option(
    "--version",
    is_flag=True,
    is_eager=True,
    expose_value=False,
    callback=print_version,
    help="Show the version and exit.",
)
def mudkip():
    """A friendly Sphinx wrapper."""


def with_application(command):
    @click.option(
        "-p",
        "--preset",
        type=click.Choice(list(Preset.registry)),
        help="Documentation preset.",
    )
    @click.option("--source-dir", type=DIRECTORY, help="The source directory.")
    @click.option("--output-dir", type=DIRECTORY, help="The output directory.")
    @click.option("--base-url", help="The url of the documentation.")
    @click.option("--repository", help="The url of the project repository.")
    @click.option("--verbose", is_flag=True, help="Show Sphinx output.")
    @wraps(command)
    def wrapper(
        preset, source_dir, output_dir, base_url, repository, verbose, *args, **kwargs
    ):
        params = dict(
            preset=preset,
            source_dir=source_dir,
            output_dir=output_dir,
            base_url=base_url,
            repository=repository,
            verbose=verbose,
        )
        for key, value in tuple(params.items()):
            if not value:
                del params[key]
        return command(*args, application=Mudkip(**params), **kwargs)

    return wrapper


@mudkip.command()
@click.option("-t", "--title", help="Documentation title.")
@with_application
def init(application, title):
    """Initialize documentation."""
    padding = "\n" * application.config.verbose

    click.secho(
        f'{padding}Initializing "{application.config.source_dir}"...', fg="cyan"
    )

    with exception_handler(exit=True, verbose=application.config.verbose):
        application.init(title)

    click.secho("\nDone.", fg="yellow")


@mudkip.command()
@click.option("-c", "--check", is_flag=True, help="Check documentation.")
@click.option(
    "--skip-broken-links",
    is_flag=True,
    help="Do not check external links for integrity.",
)
@click.option(
    "--update-gh-pages",
    is_flag=True,
    help="Update GitHub Pages.",
)
@with_application
def build(application, check, skip_broken_links, update_gh_pages):
    """Build documentation."""
    padding = "\n" * application.config.verbose

    action = (
        "Updating GitHub Pages with"
        if update_gh_pages
        else "Building and checking"
        if check
        else "Building"
    )

    click.secho(
        f'{padding}{action} "{application.config.source_dir}"...{padding}', fg="cyan"
    )

    with exception_handler(exit=True, verbose=application.config.verbose):
        application.build(
            check=check,
            skip_broken_links=skip_broken_links,
            update_gh_pages=update_gh_pages,
        )

    message = "All good" if check and not update_gh_pages else "Done"
    click.secho(f"\n{message}.", fg="yellow")


@mudkip.command()
@click.option(
    "-o",
    "--open",
    "open_browser",
    is_flag=True,
    help="Open the documentation in a browser.",
)
@click.option("--host", help="Development server host.", default="localhost")
@click.option("--port", help="Development server port.", default=5500)
@click.option("-n", "--notebook", is_flag=True, help="Open the Jupyter notebook.")
@click.option("--notebook-host", help="Notebook host.", default="localhost")
@click.option("--notebook-port", help="Notebook port.", default=8888)
@with_application
def develop(
    application, open_browser, host, port, notebook, notebook_host, notebook_port
):
    """Start development server."""
    padding = "\n" * application.config.verbose

    click.secho(
        f'{padding}Watching "{application.config.source_dir}"...{padding}', fg="cyan"
    )

    @contextmanager
    def build_manager(event_batch=None, server_url=None, notebook_url=None):
        if event_batch is None:
            if server_url:
                click.secho(f"Server running on {server_url}{padding}", fg="cyan")
            if notebook_url:
                click.secho(f"Notebook running on {notebook_url}{padding}", fg="cyan")

            with exception_handler(verbose=application.config.verbose):
                yield
            return

        now = time.strftime("%H:%M:%S")
        click.secho(f"{padding}{now}", fg="black", bold=True, nl=False)

        events = event_batch.all_events

        if len(events) == 1:
            event = events[0]
            filename = path.basename(event.src_path)
            click.echo(f" {event.event_type} {filename}{padding}")
        else:
            click.echo(f" {len(events)} changes{padding}")

        with exception_handler(verbose=application.config.verbose):
            yield

    try:
        application.develop(
            open_browser,
            host,
            port,
            notebook,
            notebook_host,
            notebook_port,
            build_manager,
        )
    except KeyboardInterrupt:
        click.secho("\nExit.", fg="yellow")


@mudkip.command()
@with_application
def test(application):
    """Test documentation."""
    padding = "\n" * application.config.verbose

    click.secho(
        f'{padding}Testing "{application.config.source_dir}"...{padding}', fg="cyan"
    )

    with exception_handler(exit=True, verbose=application.config.verbose):
        passed, summary = application.test()

    if not application.config.verbose:
        click.echo("\n" + summary)

    if passed:
        click.secho("\nPassed.", fg="yellow")
    else:
        click.secho("\nFailed.", fg="red", bold=True)
        sys.exit(1)


@mudkip.command()
@with_application
def clean(application):
    """Remove output directory."""
    padding = "\n" * application.config.verbose

    click.secho(f'{padding}Removing "{application.config.output_dir}"...', fg="cyan")

    with exception_handler(exit=True, verbose=application.config.verbose):
        application.clean()

    click.secho("\nDone.", fg="yellow")


def main():
    mudkip(prog_name="mudkip")
